js/src/vector/int.ts

```diff
@@ -17,70 +17,111 @@
 
 import { Data } from '../data';
 import { Vector } from '../vector';
+import { Chunked } from './chunked';
 import { BaseVector } from './base';
-import { VectorType as V } from '../interfaces';
-import { Int, Uint8, Uint16, Uint32, Uint64, Int8, Int16, Int32, Int64 } from '../type';
-import {
-    toInt8Array, toInt16Array, toInt32Array,
-    toUint8Array, toUint16Array, toUint32Array,
-    toBigInt64Array, toBigUint64Array
-} from '../util/buffer';
+import { VectorBuilderOptions } from './index';
+import { vectorFromValuesWithType } from './index';
+import { VectorBuilderOptionsAsync } from './index';
+import { BigInt64Array, BigUint64Array } from '../util/compat';
+import { toBigInt64Array, toBigUint64Array } from '../util/buffer';
+import { Int, Uint8, Uint16, Uint32, Uint64, Int8, Int16, Int32, Int64, IntArray } from '../type';
+import { VectorType as V, TypedArrayConstructor, BigIntArrayConstructor, BigIntArray } from '../interfaces';
+
+/** @ignore */
+type IntVectorConstructors =
+    typeof IntVector    |
+    typeof Int8Vector   |
+    typeof Int16Vector  |
+    typeof Int32Vector  |
+    typeof Uint8Vector  |
+    typeof Uint16Vector |
+    typeof Uint32Vector |
+    typeof Int64Vector  |
+    typeof Uint64Vector ;
+
+/** @ignore */
+type FromInput<T extends Int, TNull = any> =
+    IntArray | BigIntArray              |
+    Iterable<T['TValue'] | TNull>       |
+    AsyncIterable<T['TValue'] | TNull>  |
+    VectorBuilderOptions<T, TNull>      |
+    VectorBuilderOptionsAsync<T, TNull> ;
+
+/** @ignore */
+type FromArgs<T extends Int, TNull = any> = [FromInput<T, TNull>, boolean?];
+
+/** @ignore */
+type IntArrayCtor = TypedArrayConstructor<IntArray> | BigIntArrayConstructor<BigIntArray>;
 
 /** @ignore */
 export class IntVector<T extends Int = Int> extends BaseVector<T> {
 
-    public static from(this: typeof IntVector, data: Int8Array): Int8Vector;
-    public static from(this: typeof IntVector, data: Int16Array): Int16Vector;
-    public static from(this: typeof IntVector, data: Int32Array): Int32Vector;
-    public static from(this: typeof IntVector, data: Uint8Array): Uint8Vector;
-    public static from(this: typeof IntVector, data: Uint16Array): Uint16Vector;
-    public static from(this: typeof IntVector, data: Uint32Array): Uint32Vector;
+    // Guaranteed zero-copy variants
+    public static from(this: typeof IntVector, input: Int8Array): Int8Vector;
```

Comment:
Can we write this as a generic? Something like `public static from<T>(this: typeof IntVector, input: Array<T>): Vector<T>;`.

Reply:
Unfortunately not. For example, if I change the signatures to their generic form like this:
```typescript
    public static from<T extends Int>(this: typeof IntVector, input: T['TArray']): V<T>;
    public static from<T extends Int>(this: typeof IntVector, input: BigInt64Array): V<T>;
    public static from<T extends Int>(this: typeof IntVector, input: BigUint64Array): V<T>;
    public static from<T extends Int>(this: typeof IntVector, input: T['TArray'], is64bit: true): V<T>;
    public static from<T extends Int>(this: typeof IntVector, input: T['TArray'], is64bit: true): V<T>;
```

it causes the "test strong typing at compile-time" tests [here](https://github.com/apache/arrow/pull/4746/files/2a6691305ed2fb1eeff90a0286992e39fa04d967#diff-a92b06b2ed35319c47a86b2b08d0947cR231) to throw a compiler error. It seems like TS will only infer the argument type from the generic, and not the generic type from the argument:

![Screenshot from 2019-07-17 14-42-03-2](https://user-images.githubusercontent.com/178183/61413822-545b9580-a8a1-11e9-8914-f3044684ccd9.png)

Reply:
Okay, thanks for checking.